### PR TITLE
CTAP 2.3

### PR DESCRIPTION
**Pull request by 0x0ece**
See https://github.com/trussed-dev/fido-authenticator/issues/53, deps need to be modified

---

commit overview:
- outdated / dropped
  - [cargo: switch deps to ctap2.3 branches](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/1088ef7d820df880037df3411405c2d072585d00)
  - [state: add migrate_all umbrella + panic on corrupt PersistentState load](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/07ee66829e8937141d1719a550a98073bec8e10f)
- https://github.com/trussed-dev/fido-authenticator/pull/56
  - [ctap: getinfo advertises algorithms, firmwareVersion, remainingDiscov…](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/9281e4d71ea988937a1cbdf04dce47d12bb9d51c)
- https://github.com/trussed-dev/fido-authenticator/pull/57
  - [ctap2.1: implement credBlob extension (RK only, max 32 bytes)](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/33920c36b239e9ec377e2fcb120129df0ebb3fc9)
- https://github.com/trussed-dev/fido-authenticator/pull/59
  - [ctap2.1: add authenticatorConfig (0x0D) command and authnrCfg option](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/296ed0f8a1856943b933de52e6fce8166f3995cf)
  - [ctap2.1: implement setMinPINLength + minPinLength extension + forcePI…](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/c621455be6d657189f71578168d801cf1a04e6d0)
  - [ctap2.1: implement alwaysUv + toggleAlwaysUv](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/ca4531474fb6eede2f505331557b7587fbd87c86)
  - [ctap2.1: align setPin / getPinToken PIN error codes with §6.5.5](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/f452a6a2cab9b2c277966d6335f6164cfe1882bc)
  - [ctap2.1: include user field in RK allowlist GetAssertion response](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/6fe88179e60b1820e94be6c8af6b1856450b9b4b)
- https://github.com/trussed-dev/fido-authenticator/pull/62
  - [ctap2.2: implement hmac-secret-mc extension at MakeCredential time](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/60f2710d952b3e339302c005666bfd9d32a0a8c7)
- https://github.com/trussed-dev/fido-authenticator/pull/63
  - [ctap2.3: advertise smart-card transport when CCID is enabled](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/99c393b4c83ea74b127b9902aa868e569286a86f)
  - [ctap2.3: getinfo advertises FIDO_2_3](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/09858ffd1d3b747b5a0822b8f9a54b20f35f0080)
- pending
  - [ctap2.3: implement long-touch reset (Level::Strong UV, drop boot-window)](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/836e44e89292b8422de2cb07a5bfd1c165521a75)
  - [ctap2: refactor make_credential / get_assertion to take &mut Response](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/d42c1b841328c81f4d7ab52c608996fca2ae8801)
  - [ctap2.3: implement ML-DSA-44 (COSE alg -50, behind mldsa44 feature)](https://github.com/trussed-dev/fido-authenticator/pull/54/commits/29e94b39ceed9c6bd9bb9d0446ac78df4eb700a9)

### CODE REVIEW

File: src/credential.rs [hunk 1]
```diff
@@ -735,6 +749,10 @@ pub struct StrippedCredential {
     pub large_blob_key: Option<ByteArray<32>>,
     #[serde(skip_serializing_if = "Option::is_none")]
     pub third_party_payment: Option<bool>,
+    /// Carries the `credBlob` for non-resident credentials inside the
+    /// credential ID itself — see [`CredentialData::cred_blob`].
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub cred_blob: Option<Bytes<{ crate::constants::MAX_CRED_BLOB_LENGTH }>>,
```
Comment (robin-nitrokey) [inline src/credential.rs:755]: If we only support credBlob for resident keys, we don’t need the field in `StrippedCredential`, right?

File: src/lib.rs [hunk 2]
```diff
@@ -126,6 +126,10 @@ pub struct Config {
     pub large_blobs: Option<ctap2::large_blobs::Config>,
     /// Whether the authenticator supports the NFC transport.
     pub nfc_transport: bool,
+    /// Firmware version reported by `authenticatorGetInfo` (CTAP 2.1 §6.4 0x0E).
+    ///
+    /// The runner is expected to plumb its own version constant in here.
+    pub firmware_version: u32,
```
Comment (robin-nitrokey) [inline src/lib.rs:132]: I would prefer this to be an `Option<usize>` so that runners can decide to omit it and we don’t have the unnecessary conversion to `usize` in the `get_info` implementation.

File: src/constants.rs [hunk 3]
```diff
@@ -9,3 +9,8 @@ pub const ATTESTATION_CERT_ID: CertId = CertId::from_special(0);
 pub const ATTESTATION_KEY_ID: KeyId = KeyId::from_special(0);
 
 pub const MAX_RESIDENT_CREDENTIALS_GUESSTIMATE: u32 = 100;
+
+/// Maximum number of bytes of `credBlob` data we accept per credential
+/// (CTAP 2.1 §11.1; spec floor is 32). Reported in `authenticatorGetInfo`
+/// as `maxCredBlobLength`.
+pub const MAX_CRED_BLOB_LENGTH: usize = 32;
```
Comment (robin-nitrokey) [inline src/constants.rs:16]: As this is defined in `ctap-types`, the constant should also live there.

File: src/ctap2.rs [hunk 4]
```diff
@@ -570,6 +571,54 @@ impl<UP: UserPresence, T: TrussedRequirements> Authenticator for crate::Authenti
...
+
+        // CTAP 2.1 §6.11.4 step 5: a PIN/UV-auth token is required. We have no
+        // built-in UV (no biometrics), so this also implies a client PIN must
+        // be set.
+        if !self.state.persistent.pin_is_set() {
+            return Err(Error::PinNotSet);
+        }
+        let pin_protocol = request.pin_protocol.ok_or(Error::MissingParameter)?;
+        let pin_protocol = self.parse_pin_protocol(pin_protocol)?;
+        let pin_auth = request.pin_auth.ok_or(Error::MissingParameter)?;
```
Comment (robin-nitrokey) [inline src/ctap2.rs:589]: I’m a bit confused by this section:
- The referenced step 5 is just the execution of the subcommand. Typo?
- I don’t see the requirement to return `PinNotSet`. If there is no PIN, the platform was not able to obtain a PIN auth token so we handle that case anyway.
- If `pin_auth` is missing, we should return CTAP2_ERR_PUAT_REQUIRED, not CTAP2_ERR_MISSING_PARAMETER.

File: src/ctap2.rs [hunk 5]
```diff
@@ -570,6 +571,54 @@ impl<UP: UserPresence, T: TrussedRequirements> Authenticator for crate::Authenti
...
+
+        // pinUvAuthData = 0xff * 32 || 0x0d || subCommand || subCommandParams (CBOR)
+        let mut data: Bytes<{ 32 + 2 + sizes::MAX_CREDENTIAL_ID_LENGTH }> = Bytes::new();
```
Comment (robin-nitrokey) [inline src/ctap2.rs:592]: Where does this size come from?

File: src/state/migrate.rs [hunk 6]
```diff
@@ -7,6 +7,23 @@ fn ignore_does_not_exists(error: Error) -> Result<(), Error> {
...
+/// — safe to call on every boot, no-op on already-migrated state.
+///
+/// `base_path` must be the base of the file directory of the fido app
+/// (typically `fido/dat`).
+///
+/// The runner is expected to call this **before** constructing the
+/// `Authenticator`, so that any on-disk state has been brought to the latest
+/// shape before [`crate::state::PersistentState::load`] tries to deserialize
+/// it.
+pub fn migrate_all(fs: &dyn DynFilesystem, base_path: &Path) -> Result<(), Error> {
```
Comment (robin-nitrokey) [inline src/state/migrate.rs:20]: What’s the context for this function? In our implementation, migrations are triggered by the filesystem version and should not be executed on every boot.

File: src/state.rs [hunk 7]
```diff
@@ -333,12 +347,22 @@ impl PersistentState {
...
                 }
                 Err(_err) => {
-                    info!("error with previous state! {:?}", _err);
+                    // State file is present but unparseable. Silently
+                    // resetting here would silently destroy the user's PIN,
+                    // KEK, and every credential. Refuse to start instead.
+                    // The failure is at least visible in the RTT log; the
+                    // user can recover via JLink/bootloader.
+                    error_now!("PERSISTENT STATE LOAD FAILED: {:?}", _err);
+                    panic!("fido-authenticator: corrupt persistent state");
```
Comment (robin-nitrokey) [inline src/state.rs:365]: I don’t see how a panic is better than resetting the state. Realistically, the user can only fix this by performing a factory reset, right? If anything, this should be an error instead of a panic.

File: src/state.rs [hunk 8]
```diff
@@ -266,7 +266,15 @@ pub struct PersistentState {
...
+    /// PIN hash. Always serialized as a CBOR byte string (major type 2) —
+    /// the shape used since `fido-authenticator 0.2.0`. Devices upgrading
+    /// from `0.1.1` carry blobs that encode the field as a 16-element CBOR
+    /// array (major type 4); cbor-smol's `deserialize_bytes` is permissive
+    /// enough to accept both shapes (it routes major type 4 to `visit_seq`),
+    /// so the legacy format deserializes cleanly via the `serde_bytes`
+    /// helper, and the next `save` normalises the on-disk format. There is
+    /// a regression test for the 0.1.1 shape in `state::tests::deser`.
+    #[serde(default, with = "serde_bytes")]
     pin_hash: Option<[u8; 16]>,
```
Comment (robin-nitrokey) [inline src/state.rs:278]: I understand the comment describes the existing situation and not anything that is changed in this PR, right? So the actual change is adding `default`? I think that only makes sense if we also skip serializing if it is none, no?

**Review by robin-nitrokey — COMMENTED** [review]
First batch of comments. I’ve already prepared the suggested changes so you don’t have to do any changes yourself, just take a look at the comments and questions and let me know if you agree.

File: src/state.rs [hunk 9]
```diff
@@ -465,9 +487,56 @@ impl PersistentState {
...
+        // Spec: setMinPINLength may only raise the value, never lower it.
+        if new_value <= self.min_pin_length() {
+            return Err(Error::PinPolicyViolation);
+        }
+        self.min_pin_length = new_value;
+        // Spec §6.11.4 step 7: if the existing PIN is shorter than the new
+        // floor, force the platform to change it. We can't measure the
+        // existing PIN length here (only its hash is stored), so we set the
+        // flag unconditionally on any tightening.
+        if self.pin_hash.is_some() {
```
Comment (robin-nitrokey) [inline src/state.rs:516]: Why don’t we store the PIN length together with the PIN hash?

File: src/state.rs [hunk 10]
```diff
@@ -280,12 +280,34 @@ pub struct PersistentState {
...
 
 impl PersistentState {
     const RESET_RETRIES: u8 = 8;
     const FILENAME: &'static Path = path!("persistent-state.cbor");
 
+    /// Default minimum PIN length (CTAP 2.1 §6.11.4: spec floor is 4).
+    pub const DEFAULT_MIN_PIN_LENGTH: u8 = 4;
+    /// Maximum number of RP IDs that can be auto-receivers of the
+    /// `minPinLength` extension.
+    pub const MAX_MIN_PIN_LENGTH_RP_IDS: usize = 4;
```
Comment (robin-nitrokey) [inline src/state.rs:309]: Should use a constant from `ctap-types`.

File: src/ctap2.rs [hunk 11]
```diff
@@ -125,6 +127,10 @@ impl<UP: UserPresence, T: TrussedRequirements> Authenticator for crate::Authenti
         response.remaining_discoverable_credentials = remaining_discoverable_credentials
             .map(|count| count as usize);
         response.max_cred_blob_length = Some(constants::MAX_CRED_BLOB_LENGTH);
+        response.min_pin_length = Some(self.state.persistent.min_pin_length() as usize);
```
Comment (robin-nitrokey) [inline src/ctap2.rs:130]: This should probably just be a `u8` in `ctap-types` too.

File: src/state.rs [hunk 12]
```diff
@@ -465,9 +487,56 @@ impl PersistentState {
...
+        core::cmp::max(self.min_pin_length, Self::DEFAULT_MIN_PIN_LENGTH)
+    }
+
+    pub fn set_min_pin_length<T: FilesystemClient>(
+        &mut self,
+        trussed: &mut T,
+        new_value: u8,
+    ) -> Result<()> {
+        // Spec: setMinPINLength may only raise the value, never lower it.
+        if new_value <= self.min_pin_length() {
```
Comment (robin-nitrokey) [inline src/state.rs:508]: Shouldn’t this be `<` instead of `<=`?

File: src/ctap2.rs [hunk 13]
```diff
@@ -1130,6 +1134,42 @@ impl<UP: UserPresence, T: TrussedRequirements> Authenticator for crate::Authenti
 
 // impl<UP: UserPresence, T: TrussedRequirements> Authenticator for crate::Authenticator<UP, T>
 impl<UP: UserPresence, T: TrussedRequirements> crate::Authenticator<UP, T> {
+    fn config_set_min_pin_length(
```
Comment (robin-nitrokey) [inline src/ctap2.rs:1137]: Isn’t this missing the handling of the `forceChangePin` parameter?

File: src/state.rs [hunk 14]
```diff
@@ -465,9 +487,56 @@ impl PersistentState {
         pin_hash: [u8; 16],
     ) -> Result<()> {
         self.pin_hash = Some(pin_hash);
+        // Successfully (re)setting the PIN clears any pending forcePINChange
+        // request — the platform has just complied (CTAP 2.1 §6.5.5.7).
+        self.force_pin_change = false;
```
Comment (robin-nitrokey) [inline src/state.rs:492]: I think this is missing the case where the PIN is set to the old value:

> If the [forcePINChange](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#getinfo-forcepinchange) member of the [authenticatorGetInfo](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#authenticatorGetInfo) response is true and LEFT(SHA-256(newPin), 16) is equal to its internal stored LEFT(SHA-256(curPin), 16) then authenticator returns CTAP2_ERR_PIN_POLICY_VIOLATION.

**Review by robin-nitrokey — COMMENTED** [review]
I have some more questions regarding the minPinLength implementation. We should definitely have some unit tests for this feature before merging it to make sure that all edge cases are handled correctly.

File: src/state/migrate.rs (hunk 6 above)
Comment (0x0ece) [reply src/state/migrate.rs:20]: I think that this and the following panic and pin_hash default can all be deleted.

Just for your context of what I was doing...
I did an analysis of our latest firmware and what would happen to a user upgrading to this new version.
I found that there might have been some serde issues (e.g. the pin_hash), so I started creating a migrate_all fn to call all the migration steps.
But in the end there was no issue, I just had to add new fields as Option<>... so this whole thing just remained stale.

Basically we should drop the entire commit: 07ee668

File: src/state.rs (hunk 7 above)
Comment (0x0ece) [reply src/state.rs:365]: Yes agree, sorry. I marked it for myself to review later but then forgot to revert it.

File: src/state.rs (hunk 8 above)
Comment (0x0ece) [reply src/state.rs:278]: Yes, same as above. We should remove this.

File: src/state.rs [hunk 15]
```diff
@@ -266,30 +266,72 @@ pub struct PersistentState {
...
+
+    /// Load persistent state from disk.
+    ///
+    /// Returns `Ok(None)` when no state file is present (fresh / wiped
+    /// device), `Ok(Some(state))` on a successful load, and `Err(_)` when a
+    /// state file is present but cannot be deserialized. The two paths must
+    /// stay distinct so [`Self::load_if_not_initialised`] can refuse to start
+    /// the authenticator on a corrupt blob instead of silently re-initialising
+    /// (which would destroy every credential).
+    pub fn load<T: FilesystemClient>(trussed: &mut T) -> Result<Option<Self>> {
```
Comment (0x0ece) [inline src/state.rs:325]: This change to load() can also be removed, it was just to distinguish missing file from corrupted, but it the behavior is the same (wipe) it makes no difference.

File: src/state.rs (hunk 12 above)
Comment (0x0ece) [reply src/state.rs:508]: lol - could the spec be more ambiguous? 😆 
I guess if it's = you can either return err or success, the end result is the same.
I wanted to conservatively avoid any unnecessary state change.

Actually sorry, I think the spec says it should return err on <, which means it should succeed on =. So yes, we should fix.

File: src/state.rs (hunk 9 above)
Comment (0x0ece) [reply src/state.rs:516]: Personally, I would avoid it (it leaks info about the user pin).

File: src/lib.rs (hunk 2 above)
Comment (robin-nitrokey) [reply src/lib.rs:132]: Fixed in https://github.com/trussed-dev/fido-authenticator/pull/56.

File: src/ctap2.rs (hunk 4 above)
Comment (0x0ece) [reply src/ctap2.rs:589]: Yeah sorry, this is just wrong.

File: src/credential.rs (hunk 1 above)
Comment (robin-nitrokey) [reply src/credential.rs:755]: Fixed in https://github.com/trussed-dev/fido-authenticator/pull/57.

File: src/constants.rs (hunk 3 above)
Comment (robin-nitrokey) [reply src/constants.rs:16]: Fixed in https://github.com/trussed-dev/fido-authenticator/pull/57.

File: src/ctap2.rs [hunk 16]
```diff
@@ -1249,7 +1506,8 @@ impl<UP: UserPresence, T: TrussedRequirements> crate::Authenticator<UP, T> {
         //           pin.len(), pin_length, &pin);
         // chop off null bytes
         let pin_length = pin.iter().position(|&b| b == b'\0').unwrap_or(pin.len());
-        if !(4..64).contains(&pin_length) {
+        let min_pin_length = self.state.persistent.min_pin_length() as usize;
+        if pin_length < min_pin_length || pin_length >= 64 {
```
Comment (robin-nitrokey) [inline src/ctap2.rs:1510]: When working on this, we should also try to test and fix this issue:
- https://github.com/trussed-dev/fido-authenticator/issues/43

File: src/ctap2.rs (hunk 5 above)
Comment (0x0ece) [reply src/ctap2.rs:592]: Placeholder never properly cleaned up :(

Maybe we should add a const to ctap-types? This should be the precise size, but we could also add some extra just in case.
```
pub const MAX_AUTH_CONFIG_PARAMS_LENGTH: usize =
    1                                       // header
    + 3                                     // newMinPINLength
    + 2 + MAX_MIN_PIN_LENGTH_RP_IDS
        * (2 + MAX_RP_ID_LENGTH)            // minPinLengthRPIDs
    + 2;                                    // forceChangePIN
```

File: src/state/migrate.rs (hunk 6 above)
Comment (robin-nitrokey) [reply src/state/migrate.rs:20]: I see, then let’s just drop the commit and I’ll mark the relevant conversations as resolved.

File: src/state.rs (hunk 9 above)
Comment (robin-nitrokey) [reply src/state.rs:516]: I see, this is a requirement from the spec though:

> Authenticator remembers newPin length internally as PINCodePointLength.

Let’s think about it for a while and discuss next week how to deal with it.

File: src/ctap2.rs (hunk 4 above)
Comment (robin-nitrokey) [reply src/ctap2.rs:589]: Fixed in https://github.com/trussed-dev/fido-authenticator/pull/59.

File: src/state.rs (hunk 10 above)
Comment (robin-nitrokey) [reply src/state.rs:309]: Fixed in https://github.com/trussed-dev/fido-authenticator/pull/59.

File: src/ctap2.rs (hunk 11 above)
Comment (robin-nitrokey) [reply src/ctap2.rs:130]: Fixed in https://github.com/trussed-dev/fido-authenticator/pull/59.

**Comment by robin-nitrokey** [comment]
Closing in favor of https://github.com/trussed-dev/fido-authenticator/pull/60.
